[WIP] Async `run_function` with extra debugging information in case of failures

s0undt3ch opened the pull request:
### What does this PR do?
See title.

s0undt3ch committed:
Remove `linux_distribution` usage on tests for skipping

s0undt3ch committed:
Remove `runtests.py` and it's supporting code

s0undt3ch committed:
Replace `@skip_if_not_root` with `@pytest.mark.skip_if_not_root`

s0undt3ch committed:
Replace `@destructiveTest` with `@pytest.mark.destructive_test`

s0undt3ch committed:
Replace `@expensiveTest` with `@pytest.mark.expensive_test`

s0undt3ch committed:
Replace `@requires_salt_modules` with `@pytest.mark.requires_salt_modules`

s0undt3ch committed:
Replace `@skip_if_binaries_missing` with `@pytest.mark.skip_if_binaries_missing`

s0undt3ch committed:
Replace `@requires_network` with `@pytest.mark.requires_network`

s0undt3ch committed:
Use `pytest.mark.skipif` instead

s0undt3ch committed:
Skip tests if impacket or smbprotocol is not installed

s0undt3ch committed:
Switch PR's to PyTest

s0undt3ch committed:
Split the test suite runs into 'unit', 'integration', 'multimaster', by this order

On windows we don't run the multimaster tests because they weren't ran
by runtests either

s0undt3ch committed:
Move windows test white listing to pytest markers

s0undt3ch committed:
Bump PyTest requirement

dwoz commented on tests/support/case.py:
```diff
@@ -737,7 +737,7 @@ def minion_run(self, _function, *args, **kw):
         return self.run_function(_function, args, **kw)
 
-    def run_function(self, function, arg=(), minion_tgt='minion', timeout=300, master_tgt=None, **kwargs):
+    def run_function(self, function, arg=(), minion_tgt='minion', timeout=60, master_tgt=None, **kwargs):
```
Increasing this timeout drastically reduced flakiness in the test suite. I'm not sure that I am on board with changing it.

s0undt3ch replied:
Opened mostly as a test to see what breaks. And it broke a few tests.

My problem with the timeout, is not actually the timeout. It's the timeout plus the flaky decorator which defaults to 4 times. That means waiting for 20 minutes for a failure. I'm seeing that happen on the 2019.2.3 branch.

In fact this PR should have been submitted two fold, one with the timeout reduction, to see what actually needs this much time and explcitly pass a bigger timeout, the other trying the async approach.

My ultimate goal is to see if we can identify the stages of the job and act acordingly. When an assert fails we of get true is not false. So, if we could identify that, first, the job was published, second, the job was picked up by the minion and third, did the job complete or not, that would help with these unuseful assertions...

So, take this with a grain of salt, I'm not ready to argue for this cause yet, I'm still not sold on it :smile:

dwoz replied:
I agree we need to remove the flaky decorators and reduce the overall flakiness of tests. Ideally, we would not *need* such large timeouts. However, I'm not in favor of changing the timeout time unless we're sure it won't introduce the flakiness which it resolved back into the test suite.

s0undt3ch replied:
I've renamed the PR.
Right now, it's not supposed to be merged.
I'll let you know if this changes.

s0undt3ch replied:
Removed the timeout reduction changes.

s0undt3ch committed:
Hack PyTest's fixture finalization timing

s0undt3ch committed:
For `unitest.TestClass` subclasses we need to use `pytest_runtest_protocol` instead

s0undt3ch committed:
Fix cloud tests selection

s0undt3ch committed:
Only cleanup the directory if it exists

s0undt3ch committed:
This was a bad idea caught soon enough.

s0undt3ch committed:
Add `pytest-subtests` requirement

s0undt3ch committed:
Move static `norecursedirs` to `pytest.ini`

s0undt3ch committed:
Add tests to ensure proper pytest pass/fail behavior.

This is specially important because we're hacking pytest's fixture
termination and doing so in important pytest hooks which can change the
normal behavior of pytest.

s0undt3ch committed:
Temporarily skip tests which were simply ignored before pytest

s0undt3ch committed:
Attempt shorter paths for macOS

s0undt3ch committed:
These are not integration tests.

Additionally, created a base class which takes care of setup and cleanup

s0undt3ch committed:
No singleton behaviour

s0undt3ch committed:
Inherit from same base class

s0undt3ch committed:
Some improvements on Salt's global exception handler

s0undt3ch committed:
Better handling of `close()` when called from `__del__`

```
Exception AttributeError: "'AsyncReqMessageClient' object has no attribute '_closing'" in <bound method AsyncReqMessageClient.__del__ of <salt.transport.zeromq.AsyncReqMessageClient object at 0x1a0e82e90>> ignored
```

s0undt3ch committed:
Reset PyTest's `catch_log_handler` after test since pyTest doesn't do it.

s0undt3ch committed:
Actually, make it a NullHandler

s0undt3ch committed:
Only hack non `function` scoped fixtures

s0undt3ch committed:
No splits, again.

s0undt3ch committed:
Handle race condition when reaping astray processes

s0undt3ch committed:
Use `RUNTIME_VARS.FILES` instead

s0undt3ch committed:
Add `flaky` to PyTest requirements

s0undt3ch committed:
Replace `@flaky` with `@pytest.mark.flaky`

s0undt3ch committed:
Drop the old `py.test` call in favour of the new `pytest` call

s0undt3ch committed:
Allow Windows Py3 test runs to take a litle longer for now

s0undt3ch committed:
Increase start timeout for multimaster daemons

s0undt3ch committed:
Implement the `requires_salt_states` marker

s0undt3ch committed:
Reuse code instead of duplicating it

s0undt3ch committed:
Replace `@requires_salt_states` with `@pytest.mark.requires_salt_states`

s0undt3ch committed:
Add support to pass a context dictionary

s0undt3ch committed:
Implement `LoaderModuleMockMixin` as pytest fixtures

s0undt3ch committed:
Switch unit tests to use the pytest fixtures alternative to `LoaderModuleMockMixin`

s0undt3ch committed:
Switch to plain asserts

s0undt3ch committed:
Skip bad test since 0c4e38ced4a

s0undt3ch committed:
Lint disable, pylint is wrong
